analyses/doublet-detection/exploratory-analyses/02_explore-benchmark-results.Rmd

```diff
@@ -0,0 +1,267 @@
+
+
+## Setup
+
+### Packages
+
+
+```{r packages}
+suppressPackageStartupMessages({
+  library(SingleCellExperiment)
+  library(readr)
+  library(ggplot2)
+  library(caret)
+})
+
+theme_set(theme_bw())
+```
+
+### Paths
+
+
+#### Base directories
+
+```{r base paths}
+module_base <- rprojroot::find_root(rprojroot::is_renv_project)
+data_dir <- file.path(module_base, "scratch", "benchmark-datasets", params$dataset)
+result_dir <- file.path(module_base, "results", "benchmark-results")
+```
+
+#### Input and output files
+
+Set paths to input and output directories and files in the chunk below.
+
+```{r paths}
+# Input files
+scdbl_tsv <- file.path(
+  result_dir, 
+  glue::glue("{params$dataset}_scdblfinder.tsv")
+)
+scrub_tsv <- file.path(
+  result_dir, 
+  glue::glue("{params$dataset}_scrublet.tsv")
+)
+truth_tsv <- file.path(
+  data_dir, 
+  glue::glue("{params$dataset}_ground-truth.tsv")
+)
+```
+
+### Functions
+
+This chunk defines functions to explore doublet detection results
+
+
+```{r functions}
+plot_doublets <- function(df, score_col, pred_col, method_name){
+  # df is expected to contain columns `ground_truth`, `score_col`, and `pred_col`
+  # The column arguments should _not_ be provided as strings
+  ggplot(df) +
+    aes(x = ground_truth,
+        y = {{score_col}}, 
+        color = {{pred_col}}) + 
+    geom_jitter(
+      alpha = 0.2, 
+      width = 0.1
+    ) + 
+    labs(
+      x = "Ground truth call",
+      y = glue::glue("{method_name} score"),
+      color = glue::glue("{method_name} prediction")
+    )
+}
+
+
+assess_doublets <- function(df, pred_col) {
+  # df is expected to contain columns `ground_truth` and `pred_col`
+  # The `pred_col` argument _should_ be provided as a strings
+  
+  caret::confusionMatrix(
+    # truth should be first
+    table(
+      "Truth" = doublet_df$ground_truth,
+      "Prediction" = doublet_df[[pred_col]]
+    ), 
+    positive = "doublet"
+  )
+  
+}
+
+```
+
+
+### Read and prepare input data
+
+First, we'll read in and combine TSV files with doublet results.
+
+```{r}
+scdbl_df <- read_tsv(scdbl_tsv) |>
+  dplyr::select(
+    barcodes,
+    cxds_score, 
+    scdbl_score = score, 
+    scdbl_prediction  = class
+  ) |>
+  # add cxds calls at three thresholds:
+  # 0.5, 0.75, 0.9
+  dplyr::mutate(
+    cxds_prediction_0.5 = dplyr::if_else(
+      cxds_score >= 0.5,
+      "doublet",
+      "singlet"
+    ),
+    cxds_prediction_0.75 = dplyr::if_else(
+      cxds_score >= 0.75,
+      "doublet",
+      "singlet"
+    ),
+    cxds_prediction_0.9 = dplyr::if_else(
+      cxds_score >= 0.9,
+      "doublet",
+      "singlet"
+    )
+  )
+
+scrub_df <- read_tsv(scrub_tsv) 
+
+true_df <- read_tsv(truth_tsv)
+
+doublet_df <- dplyr::inner_join(
+  scdbl_df,
+  scrub_df, 
+  by = "barcodes"
+) |>
+  dplyr::inner_join(
+    true_df, 
+    by = "barcodes"
+  )
```

Comment:
Why use `inner_join` here? Theoretically every barcode should have a ground truth and this is fine, but in the case that there is no ground truth for a barcode, I think we want to know that and not drop that barcode.

Reply:
Maybe because you don't want any NAs when creating the confusion matrix, but I think we want to see those in the jitter plot. You can filter any NAs out before creating the confusion matrix?

Reply:
No real reason for this to be inner_join, will update!